2023/FerrariStanganini/NetManProject/src/main.c

```diff
@@ -0,0 +1,547 @@
+                memcpy(s_addr, &ip.ip_src.s_addr, sizeof(in_addr_t));
+                hashmap_set(hash_BH, s_addr, sizeof(in_addr_t), (uintptr_t)(void *)src_data);
+            }
+
+            if (hashmap_get(hash_BH, &ip.ip_dst.s_addr, sizeof(ip.ip_dst.s_addr), &r)) // DST not present in the hashmap
+            {
+                DATA *data = (DATA *)r;
+                data->rx_packet++;
+                if (!(data->dst)) // was only src
+                    data->dst = 1; 
+                data->time_dst = h->ts;
+                if (!(roaring_bitmap_contains(data->bitmap, ip.ip_src.s_addr)))
+                    roaring_bitmap_add(data->bitmap, ip.ip_src.s_addr);
+            }
+            else // DST not present
+            {
+                DATA *dst_data = malloc(sizeof(DATA));
+                dst_data->dst = 1;
+                dst_data->src = 0;
+                dst_data->rx_packet = 1;
+                dst_data->tx_packet = 0;
+                dst_data->time_src = h->ts;
+                dst_data->time_dst = h->ts;
+                dst_data->bitmap = roaring_bitmap_create();
+                roaring_bitmap_add(dst_data->bitmap, ip.ip_src.s_addr);
+                in_addr_t *s_addr = malloc(sizeof(in_addr_t));
+                memcpy(s_addr, &ip.ip_dst.s_addr, sizeof(in_addr_t));
+                hashmap_set(hash_BH, s_addr, sizeof(in_addr_t), (uintptr_t)(void *)dst_data);
+            }
+        }
+    }
+}
+
+/*************************************************/
+
+int main(int argc, char *argv[])
+{
+    char *device = NULL;
+    u_char c;
+    char errbuf[PCAP_ERRBUF_SIZE];
+    int promisc, snaplen = DEFAULT_SNAPLEN;
+    bitmap_BH = roaring_bitmap_create();
+    hash_BH = hashmap_create();
+
+    system("mkdir rrd_bin");
+    system("mkdir rrd_bin/db");
+    system("mkdir rrd_bin/graph");
+
+    while ((c = getopt(argc, argv, "hi:l:v:f:")) != '?')
+    {
+        if ((c == 255) || (c == (u_char)-1))
+            break;
+        switch (c)
+        {
+            case 'i':
+                device = strdup(optarg);
+                break;
+        }
+    }
+
+    if (geteuid() != 0)
+    {
+        printf("Please run this tool as superuser\n");
+        return (-1);
+    }
+
+    if (device == NULL)
+    {
+        printf("ERROR: Missing -i\n");
+        return (-1);
+    }
+
+    printf("Capturing from %s\n", device);
+    getBroadCast(device);
+    inet_pton(AF_INET, "224.0.0.0", &minMultiIP.sin_addr);
+    inet_pton(AF_INET, "239.255.255.255", &maxMultiIP.sin_addr);
+    inet_pton(AF_INET, "255.255.255.255", &allbroadcastIP.sin_addr);
+    inet_pton(AF_INET, "192.168.222.11", &intraIP.sin_addr);
+
+    promisc = 1;
+    if ((pd = pcap_open_live(device, snaplen, promisc, 500, errbuf)) == NULL)
+    {
+        printf("pcap_open_live: %s\n", errbuf);
+        return (-1);
+    }
+
+    signal(SIGINT, sigproc);
+    signal(SIGTERM, sigproc);
+    signal(SIGALRM, my_sigalarm);
```

Comment:
L'uso dell'alarm e' pericolosa perche' interrompe in modo asincrono (e quindi inaspettato mentre manipolate le strutture dati) le attività'. Va eliminato e fatto quanto volete fare periodicamente alla fine di dummyProcesssPacket (ad esempio) facendo lo scan di porzioni della hashmap e non tutte in una volta come fate voi perche' puo' portare a perdite di pacchetti